samples/electron_gui/window.js

```diff
@@ -44,103 +49,207 @@ function gui_rpc(request) {
+	if (!read_password()) {
+		return new Promise((resolve, reject)=>{
+			reject();
 		});
 	}
-	return gui_rpc("<get_auth_id/>").then(function(reply) {
+	return gui_rpc("<get_auth_id/>").then((reply)=>{
 		// TODO: error checking
-		x = new DOMParser().parseFromString(reply, "text/xml");
-		auth_id = x.getElementsByTagName("auth_id")[0].childNodes[0].nodeValue;
-		auth_salt = x.getElementsByTagName("auth_salt")[0].childNodes[0].nodeValue;
-		console.log("auth_id: "+auth_id);
-		console.log("auth_salt: "+auth_salt);
+		x = new DOMParser().parseFromString(reply, "text/xml")
+		auth_id = x.getElementsByTagName("auth_id")[0].childNodes[0].nodeValue
+		auth_salt = x.getElementsByTagName("auth_salt")[0].childNodes[0].nodeValue
+		//console.log("auth_id: "+auth_id)
+		//console.log("auth_salt: "+auth_salt)
 	});
 }
 
-// do a get_state() RPC
+// do a get_state() RPC; return promise
 //
 function get_state() {
-	return gui_rpc("<get_state/>").then(function(reply) {
-		parser = new DOMParser();
-		state = parser.parseFromString(reply, "text/xml");
+	return gui_rpc("<get_state/>").then((reply)=>{
+		parser = new DOMParser()
+		state = parser.parseFromString(reply, "text/xml")
+	})
+	.catch((err)=>{
+		show_status('disconnected')
+		console.log(err)
 	});
 }
 
 function suspend() {
-	set_run_mode("never").then(function() {
+	set_run_mode("never").then(()=>{
 	});
 }
 
+function parse_reply(reply) {
+	x = new DOMParser().parseFromString(reply, "text/xml")
+	if (x.getElementsByTagName('success')) {
+		return 0
+	}
+	return x.getElementsByTagName("status")[0].childNodes[0].nodeValue
+}
+
+function attach_poll() {
+	console.log('doing poll')
+	req = '<project_attach_poll/>'
+	gui_rpc(req).then((reply)=>{
+		console.log(reply)
+		r = parse_reply(reply)
+		switch(r) {
+		case -199:
+			setTimeout(attach_poll, 1000)
+			break
+		case 0:
+			console.log('attached')
+			break
+		default:
+			console.log('attach failed: ', reply)
+			break
+		}
+	})
+}
+
+function attach() {
+	console.log('attach')
+	url = 'https://boinc.berkeley.edu/test/'
+	auth = 'xxx'
+	req = '<project_attach>'
+	req += '<project_url>'+url+'</project_url>'
+	req += '<authenticator>'+auth+'</authenticator>'
+	req += '<project_name>BOINC test project</project_name>'
+	req += '</project_attach>'
+	gui_rpc(req).then((reply)=>{
+		console.log('did attach RPC')
+		attach_poll()
+	})
+}
+
+function client_exited(error, stdout, stderr) {
+	console.log('client exited');
+	if (error) {
+		throw error
+	}
+	console.log(stdout)
+	console.log(stderr)
+}
+
+function show_status(s) {
+	document.querySelector('#status').innerHTML = s
+}
+
+// Start the BOINC client.
+// On Win, the client will find its data directory in the registry.
+// May need to set current dir on other platforms.
+// If there's already a client running,
+// the new one will detect this and exit.
+//
+function start_client() {
+	const child = execFile(
+		'c:/Users/David/Documents/BOINC_git/boinc/win_build/Build/x64/Debug/boinc.exe',
```

Comment:
I'm afraid, this will run on your PC only. Could you please change it?

Reply:
This is demo code.  It's not intended to run anywhere other than my PC.

Reply:
But you share this code with other. It means that anyone who downloads this should be able to run it without changing manually any file

Reply:
But you share this code with other. It means that anyone who downloads this should be able to run it without changing manually any file

Reply:
It's example code, not intended to work out of the box